feat: adding prefer-theme-color-classnames rule (#22)
MetaMask/eslint-plugin-design-tokens: docs/rules/prefer-theme-color-classnames.md

# Encourage the Use of Theme Color Classnames (`@metamask/design-tokens/prefer-theme-color-classnames`)

> This rule encourages the use of theme-specific color classnames instead of hardcoded color names, promoting a centralized approach to color management via [design tokens](https://github.com/MetaMask/design-tokens). By advocating for the use of [design tokens](https://github.com/MetaMask/design-tokens) for colors, this rule helps ensure consistency, design system alignment, scalability, and ease of maintenance across your project's UI.

Author: Update to 

```suggestion
This rule encourages the use of theme color classnames instead of literal color names, promoting theme safe and consistent color management via [design tokens](https://github.com/MetaMask/design-tokens). By advocating for the use of [design tokens](https://github.com/MetaMask/design-tokens) for colors, this rule helps ensure consistency, design system alignment, scalability, and ease of maintenance across your project's UI.
```

Author: Updated here https://github.com/MetaMask/eslint-plugin-design-tokens/commit/8593ab5e6ebe6ada1833fef3b31d0de9ca317fa7

## Rule Details

> The `prefer-theme-color-classnames` rule is aimed at encouraging the use of theme-specific CSS class names over direct references to color names. This practice supports theming, reusability, and easier updates to the color palette.

Author: ```suggestion
The `prefer-theme-color-classnames` rule is aimed at encouraging the use of theme CSS class names over direct references to literal color names e.g. blue, red, green, slateGray, etc. This practice supports theming, reusability, and easier updates to the color palette.
```

Author: Updated here https://github.com/MetaMask/eslint-plugin-design-tokens/commit/8593ab5e6ebe6ada1833fef3b31d0de9ca317fa7

Examples of **incorrect** code for this rule:

```jsx
// In a JSX file
<div className="bg-slateGray-500">...</div>
<div className="text-blue-600">...</div>
```

Examples of **correct** code for this rule:

```jsx
// Using design token CSS variables

> <div className="bg-primary-default">...</div>
> <div className="text-error-default">...</div>

Author: Align theme classnames with literal colors
```suggestion
<div className="bg-default">...</div>
<div className="text-primary-default">...</div>
```

Author: Updated here https://github.com/MetaMask/eslint-plugin-design-tokens/pull/22/commits/830b1e4c586f850a53b241fae0907148dd5ce03c

```

> ## Options
> 
> This rule does not accept any options. Its primary purpose is to encourage the use of theme-specific color classnames instead of hardcoded color names.

Author: Update docs with options

```suggestion
## Options

This rule accepts a single configuration object with the following properties:

- `discouragedColors`: An array of strings representing color names that should be avoided. Providing this array will override the default set of discouraged colors. The default colors considered discouraged are:
  - `slateGray`
  - `blue`
  - `red`
  - `gray`
  - `midnight`
  - `green`
  - `orange`
  - `black`
  - `white`

### Example configuration:

```json
{
  "rules": {
    "your-plugin/prefer-theme-color-classnames": ["error", {
      "discouragedColors": ["customColor", "anotherColor"]
    }]
  }
}

```

Author: Updated here https://github.com/MetaMask/eslint-plugin-design-tokens/pull/22/commits/830b1e4c586f850a53b241fae0907148dd5ce03c

## Example Configuration

```json
{
  "@metamask/design-tokens/prefer-theme-color-classnames": "warn"
}
```

> ## When Not To Use It
> 
> You might choose not to enable this rule if:
> 
> - Your project has not adopted the MetaMask design system.
> - There are specific cases where the direct usage of hardcoded color names is necessary and cannot be replaced with themed alternatives. This includes legacy parts of your application or during a gradual migration to design tokens. In such cases, you can disable the rule for specific files or folders by adding overrides in your ESLint configuration. For example, to disable the rule in legacy files or folders, you can use:
> 
> ```json
> "overrides": [
>   {
>     "files": ["legacy/**/*.jsx", "another/path/to/specific/files/*"],
>     "rules": {
>       "@metamask/design-tokens/prefer-theme-color-classnames": "off"
>     }
>   }
> ]
> ```
> 
> This configuration allows for selective application of the rule, accommodating necessary exceptions.

Author: ```suggestion
## When Not To Use It

You might choose not to enable this rule if:

- Your project has not adopted the MetaMask design system.
- Your project does not use utility classnames such as Tailwind
- There are specific cases where the direct usage of literal color names is necessary and cannot be replaced with themed alternatives. This includes colors that should remain the same regardless of theme. In such cases, you can disable the rule for specific lines. For example, to disable the rule  for the yellow of this star icon, you can use:

```json
<Star
  className={classnames({
    // eslint-disable-next-line @metamask/design-tokens/prefer-theme-color-classnames
    'fill-yellow-500': isWatched,
    'fill-icon-muted': !isWatched,
  })}
/>
```

This configuration allows for selective application of the rule, accommodating necessary exceptions.
```

Author: Updated here https://github.com/MetaMask/eslint-plugin-design-tokens/pull/22/commits/830b1e4c586f850a53b241fae0907148dd5ce03c

## Rule Details

The `prefer-theme-color-classnames` rule is aimed at encouraging the use of CSS theme color class names over direct references to literal color names e.g. blue, red, green, slateGray, etc. This practice supports theming, reusability, and easier updates to the color palette.

Examples of **incorrect** code for this rule:

```jsx
// In a JSX file
<div className="bg-slateGray-500">...</div>
<div className="text-blue-600">...</div>
```

Examples of **correct** code for this rule:

```jsx
// Using design token theme color class names
<div className="bg-default">...</div>
<div className="text-primary-default">...</div>
```

## Options

This rule accepts an optional single configuration object with the following properties:

- `discouragedColors`: An array of strings representing color names that should be avoided. Providing this array will override the default set of discouraged colors. The default colors considered discouraged are:

>   - `slateGray`
>   - `blue`
>   - `red`
>   - `gray`
>   - `midnight`
>   - `green`
>   - `orange`
>   - `black`
>   - `white`

Author: Update list to reflect default classname list
```suggestion
- `blue`
- `red`
- `gray`
- `slateGray`
- `green`
- `orange`
- `black`
- `white`
- `indigo`
- `yellow`
- `purple`
- `pink`
- `teal`
```

Author: Updated here https://github.com/MetaMask/eslint-plugin-design-tokens/pull/22/commits/90b273ab2fd4262f02c4597684252c75b072be63

Examples of **incorrect** code for this rule:

```jsx
// In a JSX file
<div className="bg-slateGray-500">...</div>
<div className="text-blue-600">...</div>
```

Examples of **correct** code for this rule:

```jsx
// Using design token theme color class names
<div className="bg-default">...</div>
<div className="text-primary-default">...</div>
```

## Options

This rule accepts an optional single configuration object with the following properties:

- `discouragedColors`: An array of strings representing color names that should be avoided. Providing this array will override the default set of discouraged colors. The default colors considered discouraged are:
- `blue`
- `red`
- `gray`
- `slateGray`
- `green`
- `orange`
- `black`
- `white`
- `indigo`
- `yellow`
- `purple`
- `pink`
- `teal`

### Example configuration:

```json
{
  "rules": {
    "@metamask/design-tokens/prefer-theme-color-classnames": [
      "warn",
      {

>         "discouragedColors": ["customColor", "anotherColor"] // Optional

Author: Fix this

Author: Fixed https://github.com/MetaMask/eslint-plugin-design-tokens/pull/22/commits/c3afb8177174305709a9c746e971b39d4c9f68a6

      }
    ]
  }
}
```

## When Not To Use It

You might choose not to enable this rule if:

- Your project has not adopted the MetaMask design system.
- Your project does not use utility classnames such as Tailwind
- There are specific cases where the direct usage of literal color names is necessary and cannot be replaced with themed alternatives. This includes colors that should remain the same regardless of theme. In such cases, you can disable the rule for specific lines. For example, to disable the rule for the yellow of this star icon, you can use:

> ```json

Author: Update to js

Author: Fixed https://github.com/MetaMask/eslint-plugin-design-tokens/pull/22/commits/63d7faea76c04d291ca3fdd28068142b600d933b

<Star
  className={classnames({
    // eslint-disable-next-line @metamask/design-tokens/prefer-theme-color-classnames
    'fill-yellow-500': isWatched,
    'fill-icon-muted': !isWatched,
  })}
/>
```

This configuration allows for selective application of the rule, accommodating necessary exceptions.

> Adopting this rule helps ensure that your project's color definitions are centralized and easily manageable, facilitating simpler adaptations or re-theming of your application as needed.

Author: Adding comprehensive documentation about rule usage
